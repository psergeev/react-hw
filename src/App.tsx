import * as React from 'react';
import {ComponentComponent} from './components/Component.component';
import {FunctionComponent} from './components/Function.component';
import {CreateElementComponent} from './components/CreateElement.component';
import {PureComponent} from './components/Pure.component';

export class App extends React.PureComponent {
    public render() {
        return (
            <main>
                <PureComponent/>
                <ComponentComponent/>
                <FunctionComponent/>
                <CreateElementComponent/>
            </main>
        );
    }
}
