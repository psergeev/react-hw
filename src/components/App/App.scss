html, body, #app {
    font-family: Arial, "Helvetica Neue", Helvetica, sans-serif;
    margin: 0;
    min-width: 600px;
    height: 100%;
}

main {
    display: grid;
    grid-template-areas: "header"
                         "content"
                         "footer";
    grid-template-rows: auto 1fr 50px;
    grid-template-columns: 100%;
    min-height: 100%;

    header {
        grid-area: header;
    }

    section {
        grid-area: content;
        min-height: 200px;
    }

    footer {
        grid-area: footer;
        line-height: 50px;
    }
}