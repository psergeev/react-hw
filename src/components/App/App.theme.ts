import { createMuiTheme } from '@material-ui/core/styles';
import grey from '@material-ui/core/colors/grey';
import green from '@material-ui/core/colors/green';

const theme = createMuiTheme({
    palette: {
        primary: grey,
        secondary: green,
    }
});

export default theme;
