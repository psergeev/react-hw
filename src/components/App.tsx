import * as React from 'react';
import './App.scss';
import BodyWithResults from './BodyWithResults';
import Footer from './Footer';
import HeaderWithDetails from './HeaderWithDetails';
import HeaderWithSearch from './HeaderWithSearch';

export interface Movie {
    poster_path: string;
    title: string;
    tagline: string;
    release_date: string;
    vote_average: number;
    runtime: number;
    overview: string;
    genres: string[];
}

interface State {
    selectedMovie: Movie | null;
    movies: Movie[];
    hasError: boolean;
}

export class App extends React.PureComponent<{}, State> {
    public constructor(props: {}) {
        super(props);

        this.state = {
            selectedMovie: null,
            movies: [],
            hasError: false
        };

        this.handleMovieCardClick = this.handleMovieCardClick.bind(this);
        this.handleBackToSearchClick = this.handleBackToSearchClick.bind(this);
    }

    public componentDidMount() {
        this._getMovies();
    }

    public componentDidCatch() {
        this.setState({
            hasError: true,
        });
    }

    public handleMovieCardClick(title: string) {
        this.setState(prevState => ({
            selectedMovie: prevState.movies.find(movie => movie.title === title) || null
        }));
    }

    public handleBackToSearchClick() {
        this.setState({ selectedMovie: null });
    }

    public render() {
        if (this.state.hasError) {
            return <h1>Something went wrong with js code.</h1>;
        }

        return (
            <main>
                {this._getHeaderComponent()}
                <BodyWithResults
                    handleMovieCardClick={this.handleMovieCardClick}
                    genre={(this.state.selectedMovie && this.state.selectedMovie.genres.join(' & ')) as string}
                    movies={this.state.movies}
                />
                <Footer />
            </main>
        );
    }

    private _getHeaderComponent() {
        if (this.state.selectedMovie) {
            return (
                <HeaderWithDetails
                    movie={this.state.selectedMovie}
                    handleBackToSearchClick={this.handleBackToSearchClick}
                />
            );
        }

        return <HeaderWithSearch />;
    }

    private _getMovies() {
        this.setState({
            movies: [{
                poster_path: 'https://image.tmdb.org/t/p/w500/97fNAi62HawGjWru7PvVmF7RAbU.jpg',
                title: 'Kill Bill: Vol. 1',
                tagline: 'Go for the kill.',
                release_date: '2003-10-10',
                vote_average: 7.8,
                runtime: 111,
                overview: 'An assassin is shot at the altar by her ruthless employer, ' +
                    'Bill and other members of their assassination circle – but \'The Bride\' ' +
                    'lives to plot her vengeance. Setting out for some payback, she makes a death ' +
                    'list and hunts down those who wronged her, saving Bill for last.',
                genres: [
                    'Action',
                    'Crime',
                ],
            }, {
                poster_path: 'https://image.tmdb.org/t/p/w500/au9lFA5a2ZnBKCzPbZQf00r7J64.jpg',
                title: 'Kill Bill: Vol. 2',
                tagline: 'The bride is back for the final cut.',
                release_date: '2004-04-16',
                vote_average: 7.7,
                runtime: 136,
                overview: 'The Bride unwaveringly continues on her roaring rampage of revenge against ' +
                    'the band of assassins who had tried to kill her and her unborn child. She visits ' +
                    'each of her former associates one-by-one, checking off the victims on her Death List ' +
                    'Five until there\'s nothing left to do … but kill Bill.',
                genres: [
                    'Action',
                    'Crime',
                    'Thriller',
                ],
            }],
        });
    }
}
