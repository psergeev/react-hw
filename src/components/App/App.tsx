import * as React from 'react';
import './App.scss';
import BodyWithResults from '../BodyWithResults';
import Footer from '../Footer';
import HeaderWithDetails from '../HeaderWithDetails';
import HeaderWithSearch from '../HeaderWithSearch';

export interface Movie {
    id: number;
    poster_path: string;
    title: string;
    tagline: string;
    release_date: string;
    vote_average: number;
    runtime: number;
    overview: string;
    genres: string[];
}

interface Props {
    movies: Movie[];
    selectedMovie: Movie | null;
    handleMovieCardClick: (movie: Movie) => void;
}

interface AppState {
    hasError: boolean;
}

export default class extends React.PureComponent<Props, AppState> {
    public constructor(props: Props) {
        super(props);

        this.state = {
            hasError: false
        };

        this.handleMovieCardClick = this.handleMovieCardClick.bind(this);
    }

    public componentDidCatch() {
        this.setState({
            hasError: true,
        });
    }

    public handleMovieCardClick(title: string) {
        const selectedMovie = this.props.movies.find(movie => movie.title === title);

        if (selectedMovie) {
            this.props.handleMovieCardClick(selectedMovie);
        }
    }

    private _getHeaderComponent() {
        if (this.props.selectedMovie) {
            return (
                <HeaderWithDetails
                    movie={this.props.selectedMovie}
                />
            );
        }

        return <HeaderWithSearch />;
    }

    public render() {
        if (this.state.hasError) {
            return <h1>Something went wrong with js code.</h1>;
        }

        return (
            <main>
                {this._getHeaderComponent()}
                <BodyWithResults
                    handleMovieCardClick={this.handleMovieCardClick}
                    genre={(this.props.selectedMovie && this.props.selectedMovie.genres.join(' & ')) as string}
                    movies={this.props.movies}
                />
                <Footer />
            </main>
        );
    }
}
