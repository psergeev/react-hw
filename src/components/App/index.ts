import { connect } from 'react-redux';
import { showDetailsAction } from '../../state/actions';
import { State } from '../../state/reducers';
import App from './App';

const mapStateToProps = (state: State) => ({
    selectedMovie: state.selectedMovie,
    movies: state.movies
});

export default connect(
    mapStateToProps,
    {
        handleMovieCardClick: showDetailsAction
    }
)(App);
