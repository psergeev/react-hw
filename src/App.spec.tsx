import { shallow } from 'enzyme';
import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { App, Movie } from './App';

jest.mock('./components/HeaderWithDetails', () => 'HeaderWithDetails');
jest.mock('./components/HeaderWithSearch', () => 'HeaderWithSearch');
jest.mock('./components/BodyWithResults', () => 'BodyWithResults');
jest.mock('./components/Footer', () => 'Footer');

test('App component works as expected', () => {
    const component = renderer.create(
        <App />
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('App component should show header with details', () => {
    const component = shallow(<App />);
    const movies: Movie[] = component.state('movies');

    component.setState({
        selectedMovie: movies[0]
    });

    expect(component).toMatchSnapshot();
});

test('App component should go back from the description to search', () => {
    const component = shallow(<App />);
    const movies: Movie[] = component.state('movies');

    component.setState({
        selectedMovie: movies[0]
    });

    (component.instance() as App).handleBackToSearchClick();

    expect(component.find('HeaderWithSearch').length).toBe(1);
});

test('App component should show movie details on card click', () => {
    const component = shallow(<App />);
    const movies: Movie[] = component.state('movies');

    (component.instance() as App).handleMovieCardClick(movies[0].title);

    expect(component.find('HeaderWithDetails').length).toBe(1);
});

test('App component should catch js errors', () => {
    const component = shallow(<App />);

    component.setState({
        hasError: true
    });

    expect(component).toMatchSnapshot();
});

describe('First Test', () => {
    it('is working', () => {
        expect(true).toBe(true);
    });
});
