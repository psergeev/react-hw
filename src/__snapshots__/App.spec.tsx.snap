// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component should catch js errors 1`] = `
<h1>
  Something went wrong with js code.
</h1>
`;

exports[`App component should show header with details 1`] = `
<main>
  <HeaderWithDetails
    handleBackToSearchClick={[Function]}
    movie={
      Object {
        "genres": Array [
          "Action",
          "Crime",
        ],
        "overview": "An assassin is shot at the altar by her ruthless employer, Bill and other members of their assassination circle – but 'The Bride' lives to plot her vengeance. Setting out for some payback, she makes a death list and hunts down those who wronged her, saving Bill for last.",
        "poster_path": "https://image.tmdb.org/t/p/w500/97fNAi62HawGjWru7PvVmF7RAbU.jpg",
        "release_date": "2003-10-10",
        "runtime": 111,
        "tagline": "Go for the kill.",
        "title": "Kill Bill: Vol. 1",
        "vote_average": 7.8,
      }
    }
  />
  <BodyWithResults
    genre="Action & Crime"
    handleMovieCardClick={[Function]}
    movies={
      Array [
        Object {
          "genres": Array [
            "Action",
            "Crime",
          ],
          "overview": "An assassin is shot at the altar by her ruthless employer, Bill and other members of their assassination circle – but 'The Bride' lives to plot her vengeance. Setting out for some payback, she makes a death list and hunts down those who wronged her, saving Bill for last.",
          "poster_path": "https://image.tmdb.org/t/p/w500/97fNAi62HawGjWru7PvVmF7RAbU.jpg",
          "release_date": "2003-10-10",
          "runtime": 111,
          "tagline": "Go for the kill.",
          "title": "Kill Bill: Vol. 1",
          "vote_average": 7.8,
        },
        Object {
          "genres": Array [
            "Action",
            "Crime",
            "Thriller",
          ],
          "overview": "The Bride unwaveringly continues on her roaring rampage of revenge against the band of assassins who had tried to kill her and her unborn child. She visits each of her former associates one-by-one, checking off the victims on her Death List Five until there's nothing left to do … but kill Bill.",
          "poster_path": "https://image.tmdb.org/t/p/w500/au9lFA5a2ZnBKCzPbZQf00r7J64.jpg",
          "release_date": "2004-04-16",
          "runtime": 136,
          "tagline": "The bride is back for the final cut.",
          "title": "Kill Bill: Vol. 2",
          "vote_average": 7.7,
        },
      ]
    }
  />
  <Footer />
</main>
`;

exports[`App component works as expected 1`] = `
<main>
  <HeaderWithSearch />
  <BodyWithResults
    genre={null}
    handleMovieCardClick={[Function]}
    movies={
      Array [
        Object {
          "genres": Array [
            "Action",
            "Crime",
          ],
          "overview": "An assassin is shot at the altar by her ruthless employer, Bill and other members of their assassination circle – but 'The Bride' lives to plot her vengeance. Setting out for some payback, she makes a death list and hunts down those who wronged her, saving Bill for last.",
          "poster_path": "https://image.tmdb.org/t/p/w500/97fNAi62HawGjWru7PvVmF7RAbU.jpg",
          "release_date": "2003-10-10",
          "runtime": 111,
          "tagline": "Go for the kill.",
          "title": "Kill Bill: Vol. 1",
          "vote_average": 7.8,
        },
        Object {
          "genres": Array [
            "Action",
            "Crime",
            "Thriller",
          ],
          "overview": "The Bride unwaveringly continues on her roaring rampage of revenge against the band of assassins who had tried to kill her and her unborn child. She visits each of her former associates one-by-one, checking off the victims on her Death List Five until there's nothing left to do … but kill Bill.",
          "poster_path": "https://image.tmdb.org/t/p/w500/au9lFA5a2ZnBKCzPbZQf00r7J64.jpg",
          "release_date": "2004-04-16",
          "runtime": 136,
          "tagline": "The bride is back for the final cut.",
          "title": "Kill Bill: Vol. 2",
          "vote_average": 7.7,
        },
      ]
    }
  />
  <Footer />
</main>
`;
